examples/e17_message_components/src/main.rs

```diff
@@ -0,0 +1,274 @@
+use std::{
+    env,
+    error::Error as StdError,
+    fmt::{Display, Formatter, Result as FmtResult},
+    str::FromStr,
+    time::Duration,
+};
+
+use dotenv::dotenv;
+use serenity::{
+    async_trait,
+    builder::{CreateActionRow, CreateButton, CreateSelectMenu, CreateSelectMenuOption},
+    client::{Context, EventHandler},
+    futures::StreamExt,
+    model::{
+        channel::{Message, ReactionType},
+        interactions::{
+            message_component::ButtonStyle,
+            InteractionApplicationCommandCallbackDataFlags,
+            InteractionResponseType,
+        },
+    },
+    Client,
+};
+
+#[derive(Debug)]
+enum Animal {
+    Cat,
+    Dog,
+    Horse,
+    Alpaca,
+}
+
+#[derive(Debug)]
+struct ParseComponentError(String);
+
+impl Display for ParseComponentError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "Failed to parse {} as component", self.0)
+    }
+}
+
+impl StdError for ParseComponentError {}
+
+impl Display for Animal {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Self::Cat => write!(f, "Cat"),
+            Self::Dog => write!(f, "Dog"),
+            Self::Horse => write!(f, "Horse"),
+            Self::Alpaca => write!(f, "Alpaca"),
+        }
+    }
+}
+
+impl FromStr for Animal {
+    type Err = ParseComponentError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "cat" => Ok(Animal::Cat),
+            "dog" => Ok(Animal::Dog),
+            "horse" => Ok(Animal::Horse),
+            "alpaca" => Ok(Animal::Alpaca),
+            _ => Err(ParseComponentError {
+                0: s.to_string(),
+            }),
```

Comment:
While valid, it would be better to use the normal tuple constructor syntax.
```suggestion
            _ => Err(ParseComponentError(s.to_string())),
```